### PR TITLE
Implements MLFlowLogger

### PR DESCRIPTION
**Pull request by nathan-az**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
Implements a basic MLFlowLogger in-line with the functionality of the other loggers.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

----------

Note the MLFlowLogger's init was heavily inspired by the HF trainer's [MLflowCallback](https://github.com/huggingface/transformers/blob/v4.48.2/src/transformers/integrations/integration_utils.py#L1229). In-line with the other loggers, key identifiers (experiment name, run ID) can be overridden, but are often set in MLFlow via environment variables. 

Fixes https://github.com/pytorch/torchtune/issues/2211, interest shown in https://github.com/pytorch/torchtune/discussions/2103

#### Images

Logging the config shows, nested under the same path as the config output directory (this looks like it aligns to other loggers, but it's a bit ugly - do we want to keep it or should we always log in the artifact root as `torchtune_config.yaml`?)

image

Parameters and metrics show in the run overview, with nested parameters period-separated (e.g. `profiler.profile_memory`)

image

Metrics also display over-time in the Model Metrics tab according to the logging steps

image

### CODE REVIEW

**Comment by fabiogeraci** [comment]
Very similar approach but I found that self.rank pulls the local rank, prefer to log the experiment on global_rank and implement custom metrics for system log. But if the guys are happy good for me

**Comment by nathan-az** [comment]
> I found that self.rank pulls the local rank, prefer to log the experiment on global_rank and implement custom metrics for system log

I considered this, but for the sake of uniformity aligned it to the other logger implementations.

> I think you might encounter issue in setting the eun_id when running multi GPUs as well as multi nodes multi GPUs.In my custom class I also included auto_log, with true/false, for user friendly approach. But if the guys are happy good for me.

Multi-GPU should be ok due ot the `self._mlflow.active_run()` checks, and only setting the run on rank 0.

Regarding multi-node, my solution generally is to have a setup script before my trainer which sets the `MLFLOW_RUN_ID` env var on all nodes. This means there is some redundant logging on each node, but at least it aligns and doesn't create more runs.

This approach has also been useful when using MLFlow as an artifact store for multi-node runs, where different nodes have different checkpoint ranks that need to be logged together.

**Comment by fabiogeraci** [comment]
> Multi-GPU should be ok due ot the `self._mlflow.active_run()` checks, and only setting the run on rank 0.
> 
> Regarding multi-node, my solution generally is to have a setup script before my trainer which sets the `MLFLOW_RUN_ID` env var on all nodes. This means there is some redundant logging on each node, but at least it aligns and doesn't create more runs.
> 
> This approach has also been useful when using MLFlow as an artifact store for multi-node runs, where different nodes have different checkpoint ranks that need to be logged together.

Personally I prefer to let MlFlow setting the RUN_ID, simply because experiment_name and run_name can be the same and if the user forgets to change the run_id, i think MlFlow raises an error (I did not check, from memory).

I deployed my implementation on a internal server of my company, at the moment all user feed from the same front-end, therefore for me it is not an option to have redundant runs (under default), it would be confusing and messy, as well as spamming the logger database (the database team would not be happy). 

I am not a core developer therefore take this as my personal opinion

**Comment by Ankur-singh** [comment]
@lulmer cc'ing you because you request support for MLFlow logging in last office hour.

**Comment by fabiogeraci** [comment]
> @lulmer cc'ing you because you request support for MLFlow logging in last office hour.

I am not sure i understand!!

**Comment by steveepreston** [comment]
@fabiogeraci I think he mentioned you because [your message](https://github.com/pytorch/torchtune/issues/2211#issuecomment-2568064850)

**Review by joecummings — COMMENTED** [review]
This is a super quick turnaround - thanks! Can you also include some screenshots using the MLFlow logger from torchtune for additional "testing"?

**Comment by Ankur-singh** [comment]
@lulmer (Louis Ulmer) requested MLFlow logger during the office hour. Hence tagged him to this PR, so he can follow the progress. 

@fabiogeraci & @steveepreston sorry for the confusion.

**Comment by nathan-az** [comment]
No worries @joecummings - added some images! Let me know if we want to change the behaviour around the artifact naming/path. I tried to align it to the wandb logger, which just takes `output_config_fname.parent`.

My deeply nested absolute path is an unlikely case - I assume most users use relative paths for their config e.g. `configs/my_config.yaml`, which will show more cleanly.

That said, we can always just store in the artifact root `torchtune_config.yaml`. Happy to make any changes, and glad to see it's all working well 🙂

File: torchtune/training/metric_logging.py [hunk 1]
```diff
@@ -469,3 +480,116 @@ def close(self) -> None:
...
+            raise ImportError(
+                "``mlflow`` package not found. Please install mlflow using `pip install mlflow` to use MLFlowLogger."
+                "Alternatively, use the ``StdoutLogger``, which can be specified by setting metric_logger_type='stdout'."
+            ) from e
+
+        _, self.rank = get_world_size_and_rank()
+
+        self._mlflow = mlflow
+
+        self._tracking_uri = tracking_uri or os.getenv("MLFLOW_TRACKING_URI", None)
```
Comment (joecummings) [inline torchtune/training/metric_logging.py:533]: super nit: `os.getenv` should return None as the default, so no need to specify in these calls.

File: torchtune/training/metric_logging.py [hunk 2]
```diff
@@ -469,3 +480,116 @@ def close(self) -> None:
...
+        self._experiment_name = experiment_name or os.getenv(
+            "MLFLOW_EXPERIMENT_NAME", None
+        )
+        self._run_id = run_id or os.getenv("MLFLOW_RUN_ID", None)
+
+        if self.rank == 0:
+            if not self._mlflow.is_tracking_uri_set():
+                if self._tracking_uri:
```
Comment (joecummings) [inline torchtune/training/metric_logging.py:541]:
```suggestion
                if self._tracking_uri is not None:
```

File: torchtune/training/metric_logging.py [hunk 3]
```diff
@@ -469,3 +480,116 @@ def close(self) -> None:
...
+                    self._mlflow.set_tracking_uri(self._tracking_uri)
+
+            if self._mlflow.active_run() is None or self._nested_run or self._run_id:
+                if self._experiment_name:
```
Comment (joecummings) [inline torchtune/training/metric_logging.py:545]:
```suggestion
                if self._experiment_name is not None:
```

File: torchtune/training/metric_logging.py (hunk 2 above)
Comment (joecummings) [reply torchtune/training/metric_logging.py:541]: Also, what would happen if the tracking URI *is* None? Would the run fail? If so, maybe we should raise a descriptive error.

File: torchtune/training/metric_logging.py (hunk 3 above)
Comment (joecummings) [reply torchtune/training/metric_logging.py:545]: Same question here about expected behavior for a null experiment name.

File: torchtune/training/metric_logging.py [hunk 4]
```diff
@@ -469,3 +480,116 @@ def close(self) -> None:
 
     def __del__(self) -> None:
         self.close()
+
+
+class MLFlowLogger(MetricLoggerInterface):
+    """Logger for use w/ MLFlow.
```
Comment (joecummings) [inline torchtune/training/metric_logging.py:486]: Can you link to the MLFlow site here?

File: torchtune/training/metric_logging.py [hunk 5]
```diff
@@ -48,6 +48,17 @@ def save_config(config: DictConfig) -> Path:
         log.warning(f"Error saving config.\nError: \n{e}.")
 
 
+def flatten_dict(d, parent_key="", sep="."):
```
Comment (joecummings) [inline torchtune/training/metric_logging.py:51]: Please add a small comment here explaining the expected input format and output format. Also kwargs only after `d` param.

**Review by joecummings — COMMENTED** [review]
Mostly minor comments!

File: torchtune/training/metric_logging.py (hunk 5 above)
Comment (nathan-az) [reply torchtune/training/metric_logging.py]: Added a proper docstring to align to `save_config` - let me know if it's not clear!

> Also kwargs only after d param.

I wasn't exactly sure what this meant. Did you mean to explicitly use `**kwargs` then parse the dict in the function?

For now I placed `sep` before `parent_key` since `sep` is more likely to be changed by a user (although setting `parent_key` may be useful if we ever wanted to prefix the config params with `torchtune_config` or similar)

File: torchtune/training/metric_logging.py [hunk 6]
```diff
@@ -469,3 +480,116 @@ def close(self) -> None:
...
+        tracking_uri (Optional[str]): MLFlow tracking uri. If not specified, will default
+            to MLFLOW_TRACKING_URI environment variable if set, or default.
+        run_id (Optional[str]): MLFlow run name. If not specified, will default
+            to mlflow-generated HRID. Unused if run_id is specified or MLFLOW_RUN_ID
+            environment variable is found.
+        run_name (Optional[str]): MLFlow run ID. If not specified, will default
+            to MLFLOW_RUN_ID environment variable if set, or a new run will be created.
+
+    Example:
+        >>> logger = MLFlowLogger(experiment_name="my_experiment", run_name="run1", log_dir="./mlruns")
```
[the remark this answers is not in the document]
Comment (nathan-az) [reply torchtune/training/metric_logging.py]: Good catch - removed.

File: torchtune/training/metric_logging.py (hunk 3 above)
Comment (nathan-az) [reply torchtune/training/metric_logging.py:561]: If `experiment_name` isn't set, MLFlow logs runs to a `Default` experiment.

I've don't use it since I prefer the experiment name groupings, but I think it's ok not to raise an error since it's handled behaviour. That said, I'm ambivalent.

File: torchtune/training/metric_logging.py (hunk 5 above)
Comment (joecummings) [reply torchtune/training/metric_logging.py]: Ahh, sorry I wasn't being clear. In general, for any optional params we like to use the wildcard operator to force the user to specify the param using keyword only. An example is like so:

```python
def foo(a: int, *, b: Optional[int] = None) -> int:
   ...
```

This forces the user to refer to the parameter `b` in function `foo` by name:

```python
foo(1, b=2)
```

Not only does this improve readability of the code, reduce the possibility for silent errors by sending parameters in with the wrong ordering, it also makes it easier to deprecate any kwargs that we might want to change down the line. So in this case, I'd like to see:

```python
def flatten_dict(d: Dict[str, Any], *, parent_key: str = "", sep: str = ".") -> Dict[str, Any]:
    ...
```

Does that make sense?

Comment (nathan-az) [reply torchtune/training/metric_logging.py]: Ahh gotcha, love it, done.

**Review by joecummings — APPROVED** [review]
Awesome work - thanks! 🙏
